### ISSUE TITLE
remote: HTTP Basic: Access denied

### ISSUE THREAD

**Issue by zezba9000** [git-for-windows/git#1119]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

```
git version 2.12.2.windows.1
built from commit: 212247dd6345c820deeae61fcdf2f10cea10525a
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
Microsoft Windows [Version 10.0.14393]
Win10 x64
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

** not sure **

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

** git bash **

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
** git clone <http://my-GitLab-url> **
```
 - What did you expect to occur after running these commands?

** it to clone **

 - What actually happened instead?

** authentication failed **

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

** Sorry I can't **

**Comment by rappazzo** [comment]
I suspect that this is a problem in the windows credential manager.  I tried this on:
  - git 2.12.2.windows.1 => failure
  - git 2.12.1.windows.1 => success

More Details:
I have a perl script which uses (a copy of Git.pm) to invoke the credential manager.  While debugging that script, I dumped the hash that I read from the credential manager:
```
    $git->credential($cred, 'fill');
    print Data::Dumper->Dump( [ $cred ] , [ "cred" ] );
```
In 2.12.2, this produces output like this:
```
    $cred = {
      'path' => '',
      'protocol' => 'https',
      'username' => '',
      'host' => 'some.host.com',
      'password' => ''
    };
```
In 2.12.1, this produces output like this:
```
    $cred = {
      'path' => '',
      'host' => 'some.host.com',
      'protocol' => 'https',
      'password' => 'my.password',
      'username' => 'mrappazzo'
    };
```
While debugging this, I did something to get it to work on 2.12.2.  After downgrading to 2.12.1, I manually removed the credentials from Credential Manager (in Control Panel).  After successful authentication, they were back in the credential manager.  I then upgraded to 2.12.2, and I was able to successfully authenticate.

To try to recreate the problem scenario again (in 2.12.2), I cleared the credentials in Credential Manager.  Reattempting to authenticate gave the credentials prompt.  The output of the perl hash was missing the password again (thus, reproducing the error condition).

**Comment by linquize** [comment]
Do you use `credential.helper manager` (Microsoft) or `credential.helper wincred` (Legacy)?

**Comment by rappazzo** [comment]
Not sure if you are asking me, but this is my answer: 
```
$ git config credential.helper
manager
```

**Comment by linquize** [comment]
Can you temporarily change to `wincred` and then try http basic auth?

**Comment by rappazzo** [comment]
Maybe it is the connection to the password dialog popup.  It even seems to fail with no credential manager:
```
mrappazzo@ MINGW64 /d/code/project (master)
$ git config --global credential.helper manager

mrappazzo@ MINGW64 /d/code/project (master)
$ git fetch origin
fatal: Authentication failed for 'https://url.com/project.git/'

mrappazzo@ MINGW64 /d/code/project (master)
$ git config --global credential.helper wincred

mrappazzo@ MINGW64 /d/code/project (master)
$ git fetch origin
fatal: Authentication failed for 'https://url.com/project.git/'

mrappazzo@ MINGW64 /d/code/project (master)
$ git config --global --unset credential.helper

mrappazzo@ MINGW64 /d/code/project (master)
$ git fetch origin
fatal: Authentication failed for 'https://url.com/project.git/'
```

**Comment by whoisj** [comment]
This is most likely a [GCM bug](https://github.com/Microsoft/Git-Credential-Manager-for-Windows/issues/399).

**Comment by dscho** [comment]
> If the problem was occurring with a specific repository, can you provide the
>     URL to that repository to help us with testing?
> 
> ** Sorry I can't **

I wish it were clear to everybody writing such an answer how difficult it makes it to try to help... :sad:

**Comment by zezba9000** [comment]
Well I can't share my companies source code.
I can make a reproducible repo if needed? Although I think this happens on any repo if you have a clean install of Windows.

**Comment by dscho** [comment]
> I can make a reproducible repo if needed?

Umm. If you want me to help you, it would be a good idea to make that as easy as you can. So yes, I think that a reproducible repro is needed.

> Although I think this happens on any repo if you have a clean install of Windows.

The problem is not the repository. The problem is he setup where you talk to a server via Basic HTTP authentication. If you have such a server available, or if you make it easy to set one up (e.g. by writing a simple node.js script that starts a simple web server with Basic  HTTP authentication), it constitutes an MCVE, as it is minimal, it is complete (because the point is not to fetch, but to authenticate successfully), it is verifiable (because it can be started locally).

> So VirtualBox might help here.

I run Windows on my machine.

In any case, @whoisj identified the problem based on other bug reports and fixed it. As soon as the executables are code-signed, we will have a new Git Credential Manager version that resolves this ticket.

**Comment by zezba9000** [comment]
I really think you had a contextual misunderstanding of what I said. I feel like you're putting a bunch of baggage on what the chain of events was here.

> Umm. If you want me to help you, it would be a good idea to make that as easy as you can.... The problem is not the repository. The problem is he setup where you talk to a server via Basic HTTP authentication.

Ok umm nothing... I had no idea the issue was related to HTTP directly, for all I know the Windows credential manager was messed up and wasn't passing back the GUI's user entries back correctly to git. So if you don't have a HTTP url to test with, thats fine, just ask. I can probably set up some port forwarded URL from home you can use. I don't think its a good idea for me to dish out my works IP addresses for this... but as you stated you found the issue in the cred manager. So are you still asking for the GitLab HTTP url to test with?

> I run Windows on my machine.

This was not the point (of course you do, this is **Git for Windows** after all). Because this issue doesn't happen in all conditions with 2.12.2, if say you clone with 2.11.1 then update and clone again (if I remember correctly), the issue doesn't happen. If you however have a clean install of Windows on VirtualBox say and only install 2.12.2 you will always hit the issue... was just giving an example of how to "always" reproduce the issue.

**Comment by ViceIce** [comment]
For me it looks like #1118. But there the version 2.12.1 is broken too, because of the new Git Credential Manager v1.9.0.

When the credentials are stored everything is working fine. If they are missing the credentials dialog is not able to pass the credentials to git correctly.

**Comment by dscho** [comment]
I am fairly certain that this issue is fixed with Git Credential Manager v1.9.1, which will be bundled with Git for Windows v2.12.2(2).

**Comment by thargenediad** [comment]
On a Windows Server 2016 Datacenter VM in Azure, I had to use wincred, to get authentication to work with my company's TFS 2017.

`git config --global credential.helper wincred`
